src/NHibernate.Test/Linq/ParameterTests.cs

```diff
@@ -0,0 +1,120 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using System.Text.RegularExpressions;
+using NUnit.Framework;
+
+namespace NHibernate.Test.Linq
+{
+	[TestFixture]
+	public class ParameterTests : LinqTestCase
+	{
+		[Test]
+		public void UsingSameArrayParameterTwice()
+		{
+			var ids = new[] {11008, 11019, 11039};
+			AssertTotalParameters(
+				db.Orders.Where(o => ids.Contains(o.OrderId) && ids.Contains(o.OrderId)),
+				ids.Length);
+		}
+
+		[Test]
+		public void UsingDifferentArrayParameters()
+		{
+			var ids = new[] { 11008, 11019, 11039 };
+			var ids2 = new[] { 11008, 11019, 11039 };
+			AssertTotalParameters(
+				db.Orders.Where(o => ids.Contains(o.OrderId) && ids2.Contains(o.OrderId)),
+				ids.Length + ids2.Length);
+		}
+
+		[Test]
+		public void UsingSameListParameterTwice()
+		{
+			var ids = new List<int> { 11008, 11019, 11039 };
+			AssertTotalParameters(
+				db.Orders.Where(o => ids.Contains(o.OrderId) && ids.Contains(o.OrderId)),
+				ids.Count);
+		}
+
+		[Test]
+		public void UsingDifferentListParameters()
+		{
+			var ids = new List<int> { 11008, 11019, 11039 };
+			var ids2 = new List<int> { 11008, 11019, 11039 };
+			AssertTotalParameters(
+				db.Orders.Where(o => ids.Contains(o.OrderId) && ids2.Contains(o.OrderId)),
+				ids.Count + ids2.Count);
+		}
+
+		[Test]
+		public void UsingSameEntityParameterTwice()
+		{
+			var order = db.Orders.First();
+			AssertTotalParameters(
+				db.Orders.Where(o => o == order && o != order),
+				1);
+		}
+
+		[Test]
+		public void UsingDifferentEntityParameters()
+		{
+			var order = db.Orders.First();
+			var order2 = db.Orders.Skip(1).First();
+			AssertTotalParameters(
+				db.Orders.Where(o => o == order && o != order2),
+				2);
+		}
+
+		[Test]
+		public void UsingSameValueTypeParameterTwice()
+		{
+			var value = 1;
+			AssertTotalParameters(
+				db.Orders.Where(o => o.OrderId == value && o.OrderId != value),
+				1);
+		}
+
+		[Test]
+		public void UsingDifferentValueTypeParameters()
+		{
+			var value = 1;
+			var value2 = 2;
```

Comment:
What about case `var value2 = 1;`? Will it be still 2  parameters or 1? I think it should still be 2 parameters (at least by default). As this query modifications due to parameter values may affect performance on DBs with cached execution plan.

Reply:
> What about case var value2 = 1;? Will it still be 2 parameters or 1?

In the current state of this PR, only one parameter would be generated. Please check my last [comment](https://github.com/nhibernate/nhibernate-core/pull/2335#issuecomment-602035664).